crates/sui-network/src/state_sync/mod.rs

```diff
@@ -0,0 +1,854 @@
+    peer_id: PeerId,
+    response: &Result<Option<Checkpoint>, Status>,
+) {
+    match response {
+        Ok(latest) => {
+            peer_heights
+                .write()
+                .unwrap()
+                .update_peer_height(peer_id, latest.clone());
+        }
+        Err(status) => {
+            trace!("get_latest_checkpoint_summary request failed: {status:?}");
+            peer_heights.write().unwrap().heights.remove(&peer_id);
+        }
+    }
+}
+
+async fn query_peers_for_their_latest_checkpoint(
+    network: anemo::Network,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    tx: mpsc::WeakSender<StateSyncMessage>,
+) {
+    let peer_heights = &peer_heights;
+    let futs = peer_heights
+        .read()
+        .unwrap()
+        .heights
+        .keys()
+        // Filter out any peers who we aren't connected with
+        .flat_map(|peer_id| network.peer(*peer_id))
+        .map(|peer| {
+            let peer_id = peer.peer_id();
+            let mut client = StateSyncClient::new(peer);
+
+            let request =
+                Request::new(GetCheckpointSummaryRequest::Latest).with_timeout(DEFAULT_TIMEOUT);
+            async move {
+                let response = client
+                    .get_checkpoint_summary(request)
+                    .await
+                    .map(Response::into_inner);
+                update_peer_height(peer_heights, peer_id, &response);
+                response
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let checkpoints = futures::future::join_all(futs)
+        .await
+        .into_iter()
+        .flatten()
+        .flatten();
+
+    let highest_checkpoint = checkpoints.max_by_key(|checkpoint| checkpoint.sequence_number());
+
+    let our_heighest_checkpoint = peer_heights
+        .read()
+        .unwrap()
+        .highest_known_checkpoint()
+        .cloned();
+
+    let _new_checkpoint = match (highest_checkpoint, our_heighest_checkpoint) {
+        (Some(theirs), None) => theirs,
+        (Some(theirs), Some(ours)) if theirs.sequence_number() > ours.sequence_number() => theirs,
+        _ => return,
+    };
+
+    if let Some(tx) = tx.upgrade() {
+        let _ = tx.send(StateSyncMessage::StartSyncJob).await;
+    }
+}
+
+async fn sync_to_checkpoint<S: WriteStore>(
+    network: anemo::Network,
+    store: S,
+    peer_heights: Arc<RwLock<PeerHeights>>,
+    checkpoint_header_download_concurrency: usize,
+    checkpoint: Checkpoint,
+) -> Result<()> {
+    let mut current = store.get_highest_trusted_checkpoint();
+    if current.as_ref().map(|x| x.sequence_number()) >= Some(checkpoint.sequence_number()) {
+        return Err(anyhow::anyhow!(
+            "target checkpoint {} is older than highest trusted checkpoint {}",
+            checkpoint.sequence_number(),
+            current.map(|x| x.sequence_number()).unwrap_or(0)
+        ));
+    }
+
+    let mut rng = <rand::rngs::StdRng as rand::SeedableRng>::from_entropy();
```

Comment:
again wondering if thread_rng would be suitable here?

Reply:
a thread_rng can't be held across an await point and presently the two places rngs are used, are written expecting that it can be held across await points.